File: availability/simple-availability-swift/Sources/SimpleAvailability/Application/OverdueLockHandling.swift
import Foundation

struct OverdueLockHandling {
    let availabilityService: AvailabilityService
    let repository: AssetAvailabilityRepository

    func unlockOverdue() {
        for item in repository.findOverdue() {
            availabilityService.unlockIfOverdue(assetAvailability: item)
        }
    }
}

File: availability/simple-availability/src/main/java/com/softwarearchetypes/availability/application/OverdueLockHandling.java
package com.softwarearchetypes.availability.application;

import org.springframework.scheduling.annotation.Scheduled;

import com.softwarearchetypes.availability.domain.AssetAvailabilityRepository;

class OverdueLockHandling {

    private final AvailabilityService availabilityService;
    private final AssetAvailabilityRepository repository;

    OverdueLockHandling(AvailabilityService availabilityService, AssetAvailabilityRepository repository) {
        this.availabilityService = availabilityService;
        this.repository = repository;
    }

    @Scheduled(fixedDelay = 10)
    void unlockOverdue() {
        repository
                .findOverdue()
                .forEach(availabilityService::unlockIfOverdue);
    }
}
